File: 30-Day LeetCoding Challenge/C++/FirstUnique.cpp
class FirstUnique {
public:
    unordered_map<int, int> count_in_q;
    queue<int> q;
    FirstUnique(vector<int>& nums) {
        for(int num : nums) {
            q.push(num);
            count_in_q[num]++;
        }
    }
    
    int showFirstUnique() {
        while(!q.empty()) {
            int first = q.front();
            if(count_in_q[first] == 1) {
                return first;
            }
            else {
                q.pop();
            }
        }
        return -1;
    }
    
    void add(int value) {
        q.push(value);
        count_in_q[value]++;
    }
};

/**
 * Your FirstUnique object will be instantiated and called as such:
 * FirstUnique* obj = new FirstUnique(nums);
 * int param_1 = obj->showFirstUnique();
 * obj->add(value);
 */

File: 30-Day LeetCoding Challenge/Python/FirstUnique.py
class FirstUnique(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.q = []
        self.count_in_q = {}
        for num in nums:
            self.q.append(num)
            if num in self.count_in_q:
                self.count_in_q[num] += 1
            else:
                self.count_in_q[num] = 1
        

    def showFirstUnique(self):
        """
        :rtype: int
        """
        while self.q:
            first = self.q[0]
            if (self.count_in_q[first] == 1):
                return first
            else:
                self.q.pop(0)
        return -1

    def add(self, value):
        """
        :type value: int
        :rtype: None
        """
        self.q.append(value)
        if value in self.count_in_q:
            self.count_in_q[value] += 1
        else:
            self.count_in_q[value] = 1


# Your FirstUnique object will be instantiated and called as such:
# obj = FirstUnique(nums)
# param_1 = obj.showFirstUnique()
# obj.add(value)